augur/filter/__init__.py

```diff
@@ -68,8 +68,32 @@ def register_arguments(parser):
     subsample_limits_group.add_argument('--sequences-per-group', type=int, help="subsample to no more than this number of sequences per category")
     subsample_limits_group.add_argument('--subsample-max-sequences', type=int, help="subsample to no more than this number of sequences; can be used without the group_by argument")
     group_size_options = subsample_group.add_mutually_exclusive_group()
-    group_size_options.add_argument('--probabilistic-sampling', action='store_true', help="Allow probabilistic sampling during subsampling. This is useful when there are more groups than requested sequences. This option only applies when `--subsample-max-sequences` is provided.")
+    group_size_options.add_argument('--probabilistic-sampling', action='store_true', help="Allow probabilistic sampling during subsampling. This is useful when there are more groups than requested sequences. This option only applies when `--subsample-max-sequences` is provided. Cannot be used with `--group-by-weights`.")
     group_size_options.add_argument('--no-probabilistic-sampling', action='store_false', dest='probabilistic_sampling')
+    group_size_options.add_argument('--group-by-weights', type=str, metavar="FILE", help="""
+        TSV file defining weights for grouping. Requirements:
+
+        (1) The first row must be a header.
+        (2) There must be a numeric ``weight`` column (weights can take on any
+            non-negative values).
+        (3) Other columns must be a subset of columns used in ``--group-by``,
+            with combinations of values covering all combinations present in the
+            metadata.
+        (4) This option only applies when ``--group-by`` and
+            ``--subsample-max-sequences`` are provided.
+        (5) This option cannot be used with ``--probabilistic-sampling``.
```

Comment:
I believe that `--probabilistic-sampling` should be compatible with `--group-by-weights`. 

Here is a quick example. Let's say we have countries `A`, `B`, `C`, etc...,  through `J`. We specify `--subsample-max-sequences of 5`. With 10 groups and with 5 total sequences targeted you should get `--sequences-per-group` of 0.5. Without weighted subsampling this would target `A` with 0.5 total sequences, `B` with 0.5 total sequences, etc... 

Now let's say we specify `--group-by-weights weights.tsv` with
```
country weight
A       10
B       2
C       1
D       1
E       1
F       1
G       1
H       1
I       1
J       1
```

With above weights you grab `A` 10/20 = 50% of the time when you grab a sample, you grab `B` 2/20 = 10% of the time when you grab a sample, you grab `C` 1/20 = 5% of the time when you grab a sample, etc...

So, if you specify `--subsample-max-sequences 5` the final targets should be, `A` at 50% × 5 = 2.5 total sequences probabilistically sampled, `B` at 10% × 5 = 0.5 total sequences probabilistically sampled, `C` at 5% × 5 = 0.25 total sequences probabilistically sampled, etc...

This same logic should extend to more granular groups created by `--group-by country year`, etc...

I think it's important that probabilistic sampling eventually work with `--group-by-weights`. However, I'm very okay to merge this PR without support for `--probabilistic-sampling` and follow up soon afterwards with a separate PR that implements this.

Reply:
Good point to raise. Probabilistic sampling is inherent to the weighted sampling implementation:

https://github.com/nextstrain/augur/blob/66d0df4b9c7f5396a7f7e90a7c7e6b08d96a7d2e/augur/filter/subsample.py#L449-L454

<details>
<summary>
Detailed example
</summary>

```sh
cat >metadata.tsv <<~~
strain	country
A1	A
B1	B
C1	C
D1	D
E1	E
F1	F
G1	G
H1	H
I1	I
J1	J
~~

cat >weights.tsv <<~~
country	weight
A	10
B	2
C	1
D	1
E	1
F	1
G	1
H	1
I	1
J	1
~~

augur filter \
  --metadata metadata.tsv \
  --group-by country \
  --group-by-weights weights.tsv \
  --subsample-max-sequences 5 \
  --subsample-seed 0 \
  --output-group-by-sizes target_group_sizes.tsv \
  --output-metadata filtered.tsv
```

Output:

```
WARNING: Asked to provide at most 5 sequences, but there are 10 groups.
Sampling with weights defined by weights.tsv.
WARNING: Targeted 3 sequences for group ["country='A'"] but only 1 is available.
6 strains were dropped during filtering
        6 were dropped because of subsampling criteria
4 strains passed all filters
```
`target_group_sizes.tsv` looks like:

```
cat target_group_sizes.tsv | tsv-pretty 
country  weight  _augur_filter_target_size  _augur_filter_input_size  _augur_filter_output_size
A            10                          3                         1                          1
B             2                          0                         1                          0
C             1                          0                         1                          0
D             1                          0                         1                          0
E             1                          1                         1                          1
F             1                          1                         1                          1
G             1                          0                         1                          0
H             1                          0                         1                          0
I             1                          0                         1                          0
J             1                          1                         1                          1
```

</details>


However, it uses a different approach than [get_probabilistic_group_sizes](https://github.com/nextstrain/augur/blob/d8faf01ec1008b4e7b98fa2285d2f9c931c505d7/augur/filter/subsample.py#L252) which instead samples from a Poisson distribution. I think it might be possible to replace that with the approach I used here, but I'll investigate that separately.

TLDR: I've worded this poorly and probabilistic sampling is always used. I'll update it to say "this option cannot be used with ``--no-probabilistic-sampling``".

Reply:
Ah! Awesome. This makes sense. I just tested this exact scenario and everything works as I expected with "probabilistic sampling". If you could update the help text as you say, this would be a bit less confusing.

Reply:
Help text updated:

https://github.com/nextstrain/augur/blob/8e3d6be81559a25c41ce1d98342a5ac458bffdd7/augur/filter/__init__.py#L85

Good that you caught this. It required some additional tweaks and proper validation:

https://github.com/nextstrain/augur/blob/8e3d6be81559a25c41ce1d98342a5ac458bffdd7/augur/filter/validate_arguments.py#L61-L65